metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -1455,30 +1524,56 @@ private void replay(ApiMessage message, Optional<OffsetAndEpoch> snapshotId, lon
+            Set<SnapshotReason> snapshotReasons = new HashSet<>();
+            // Check if a snapshot should be generated because of committed bytes
+            if (committedBytesSinceLastSnapshot >= snapshotMaxNewRecordBytes) {
+                snapshotReasons.add(
+                    SnapshotReason.maxBytesExceeded(committedBytesSinceLastSnapshot, snapshotMaxNewRecordBytes)
+                );
             }
 
-            log.info(
-                "Generating a snapshot that includes (epoch={}, offset={}) after {} committed bytes since the last snapshot because, {}.",
-                lastCommittedEpoch,
-                lastCommittedOffset,
-                committedBytesSinceLastSnapshot,
-                SnapshotReason.MaxBytesExceeded
-            );
+            // Check if a snapshot should be generated because of committed append times
+            if (snapshotMaxIntervalMs > 0) {
+                // Time base snasphots are enabled
+                long snapshotIntervalMs = time.milliseconds() - oldestCommittedLogOnlyAppendTimestamp;
+                if (snapshotIntervalMs >= snapshotMaxIntervalMs) {
+                    snapshotReasons.add(SnapshotReason.maxIntervalExceeded(snapshotIntervalMs, snapshotMaxIntervalMs));
+                } else {
+                    maybeScheduleNextGenerateSnapshot();
+                }
+            }
 
-            snapshotGeneratorManager.createSnapshotGenerator(
-                lastCommittedOffset,
-                lastCommittedEpoch,
-                lastCommittedTimestamp
-            );
-            committedBytesSinceLastSnapshot = 0;
+            if (!snapshotReasons.isEmpty()) {
+                if (!isActiveController()) {
+                    // The inactive controllers only create an in-memory snapshot when generating a snapshot. This is
+                    // unlike the active controller which creates in-memory snapshots every time an uncommitted batch
+                    // gets replayed.
+                    snapshotRegistry.getOrCreateSnapshot(lastCommittedOffset);
+                }
+
+                log.info(
+                    "Generating a snapshot that includes (epoch={}, offset={}) because: {}",
+                    lastCommittedEpoch,
+                    lastCommittedOffset,
+                    SnapshotReason.stringFromReasons(snapshotReasons)
+                );
+
+                snapshotGeneratorManager.createSnapshotGenerator(lastCommittedOffset, lastCommittedEpoch, lastCommittedTimestamp);
+
+                // Reset all of the snapshot counters
+                committedBytesSinceLastSnapshot = 0;
+                oldestCommittedLogOnlyAppendTimestamp = Long.MAX_VALUE;
+
+                // Starting a snapshot invalidates any scheduled snapshot generation
+                cancelNextGenerateSnapshot();
+            }
+        } else {
+            /* Skip snapshot generation if there is a snaphshot in progress.
```

Comment:
nit: I think this would be a little clearer if the `if` check is inverted:
```java
if (snapshotGeneratorManager.snapshotInProgress()) {
  /* Skip snapshot generation if there is a snaphshot in progress.
...
} else {
```